srml/contract/src/account_db.rs

```diff
@@ -83,23 +140,24 @@ impl<T: Trait> AccountDb<T> for DirectAccountDb {
 			}
 			for (k, v) in changed.storage.into_iter() {
 				if let Some(value) = v {
-					<StorageOf<T>>::insert(&address, &k, value);
+					child::put_raw(&keyspace[..], &k, &value[..]);
 				} else {
-					<StorageOf<T>>::remove(&address, &k);
+					child::kill(&keyspace[..], &k);
 				}
 			}
 		}
 	}
 }
-
 pub struct OverlayAccountDb<'a, T: Trait + 'a> {
 	local: RefCell<ChangeSet<T>>,
+	keyspace_account: Option<RefCell<AccountKeySpaceMapping<<T as system::Trait>::AccountId>>>,
```

Comment:
Why do we need this kind of chaining of `AccountKeySpaceMapping`? Can we simplify it, e.g. by using a single instance of it?

Reply:
I firstly wrote the code with a single instance under a Rc RefCell, but I did not like the way caching would be done at every overlay layer, but it is only a matter of specializing the first layer for being the only one doing the caching (it would make the code more complex but I can still try switch back to that).

Reply:
> but I did not like the way caching would be done at every overlay layer

But now, it looks like it also done on each layer, no? The cache is modified for the each layer (i.e. grep `ka_mut.insert` or `ka.borrow_mut().insert`)?

Reply:
Ah, ok, I see now, everything is wrapped in a `map` and `keyspace_account` is `None` for the every layer except for the second (or in the other words, the first `OverlayAccountDb` layer).

Well, tbh, I think this is rather convoluted from the code perspective, and it sounds like a single cache instance should be a more simpler solution, no? why no?

Reply:
It also need to have this kind of specialization otherwise insert of value from NativeAccount will be done at every OverlayAccount (rc will work super nice for access, not for insert). But it is only adding a bool at the first layer to ensure the update is only done under this condition, I can switch to it (there is an interest in term of perf).